Change API to remove global parameters

ArteMisc opened the pull request:
These commits aim to reduce the reliance on global state, and bring the API closer to the ones found in Golang's standard library (such as the crypto.Signer/crypto.SignerOpts interfaces).

Includes some more lint-related changes such as docs that follow the format of
```golang
// SomeFunc does...
func SomeFunc() { }
```

Commit:
update api, remove global state

Commit:
merge

Commit:
add install doc

Commit:
updated constants

Commit:
putting these methods on hasher makes more sense

lentus commented on const.go:
```diff
@@ -0,0 +1,16 @@
+
+// constants for WOTPS signatures for both W4 and W16 modes.
+const (
+	W4PublicKeyBytes = 2144        // size of publis key
```
in comment: publis -> public

lentus replied:
W4PublicKeyBytes should be 4256

lentus commented on const.go:
```diff
@@ -0,0 +1,16 @@
+// constants for WOTPS signatures for both W4 and W16 modes.
+const (
+	W4PublicKeyBytes = 2144        // size of publis key
+	W4SecretKeyBytes = W4SeedBytes // TODO should this be the size of the internal private key, or the seed?
```
W4SecretKeyBytes is redundant because the secret key is never exposed: it is recomputed (using expandSeed()) when required to save storage. The "secret key" for WOTS+ is thus actually the secret seed, which is always 32 bytes long.

ArteMisc replied:
These constant names were inspired by the equivalent ones found in Godium, where ed25519 has a private key that contains both seed and public key, but not the "private key" as used in the algorithm, which seems similar to the situation in wotsp. It is in that sense merely used for documentation purposes, to show that the seed "is the secret" from the library's user's point of view.

lentus commented on const.go:
```diff
@@ -0,0 +1,16 @@
+const (
+	W4PublicKeyBytes = 2144        // size of publis key
+	W4SecretKeyBytes = W4SeedBytes // TODO should this be the size of the internal private key, or the seed?
+	W4Bytes          = 2144        // size of signatures
```
Signature size if 4256 bytes (same as the public key)

lentus commented on const.go:
```diff
@@ -0,0 +1,16 @@
+	W4PubSeedBytes   = 32
+
+	W16PublicKeyBytes = 2144
+	W16SecretKeyBytes = W16SeedBytes
```
Again, redundant

lentus commented on const.go:
```diff
@@ -0,0 +1,16 @@
+
+	W16PublicKeyBytes = 2144
+	W16SecretKeyBytes = W16SeedBytes
+	W16Bytes          = 4256
```
Should be 2144, same as public key size

lentus commented on params.go:
```diff
@@ -0,0 +1,33 @@
+// Opts implements crypto.SignerOpts and crypto.DecrypterOpts.
+type Opts struct {
+	Mode    Mode
+	PubSeed []byte
```
PubSeed is always required to be a slice of bytes of length 32, for pubkey gen, signing and verification. To reflect this I think it should be an explicit parameter of those functions. Also, this allows the use of Opts{} in those functions for the default values (Mode = W16 and Address an all-zero address).

ArteMisc replied:
Sounds reasonable

ArteMisc replied:
another consideration is to accept `*Opts` instead of `Opts` to allow passing `nil` as valid options. Opts could then implement methods to decide mode/address values, like such:

```golang
func (o *Opts) mode() Mode {
    if o == nil {
        return W16
    }
    return o.Mode
}
```

lentus commented on params.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// mode should implement crypto.SignerOpts
+var _ crypto.SignerOpts = Opts{}
```
WOTS+ should not be used as is: it is meant as a building block for hash-based signature schemes such as XMSS and SPHINCS. The purpose of this repo is to provide a standalone implementation that can be used in larger constructions, which is what I needed when doing my Master's thesis (I will add this purpose to the README later). Because it should not be used as is, I don't think this implementation should implement the Signer interface, and thus Opts does not need to implement SignerOpts either.

ArteMisc replied:
Maybe it's worth making it an /internal/wotsp/ part of a post-quantum crypto library, as opposed to a separate library. Otherwise one could argue that even if used as a one-time only Signer, it is still a Signer, and it could use the interface implementation.

That said, it seems the /x/crypto/ packages don't tend to implement the standard library interfaces either.

lentus commented on mode.go:
```diff
@@ -0,0 +1,61 @@
+	"fmt"
+)
+
+// Mode specifies the mode in which
```
This line can be removed

ArteMisc replied:
whoops

lentus reviewed (changes requested):
Looks good, some changes required to the constants and Opts does not need to implement SignerOpts

ArteMisc committed:
update consts

fixed incorrect constat values, updated comments, added const for size of Address

ArteMisc committed:
cleanup mode.go
